### ISSUE TITLE
Reexport context related apis from react-admin isntead of relying on 3rd parties like react-router or react-hook-form

### ISSUE THREAD

**Issue by crtl** [marmelab/react-admin#8481]
**Is your feature request related to a problem? Please describe.**
I am currently developing an admin dashboard for a running frontend app.
The frontend app already uses react-router, react-router-dom and react-hook-form with specific versions.
react-admin uses its own versions of these libs which then causes conflicts in my apps.
Everything works well until I import anything from these conflicting libraries.

**Describe the solution you'd like**
react-admin should re export all the hooks and components required to use it by itself (Link, useHistory, useForm etc.). 

**Describe alternatives you've considered**
Removing my project dependencie and relying on react-admin dependencies but this is a pretty dirty workaround.
Also I have to adapt my whole project to match the 3rd party dependency requirements of react-admin which is not how modules should work. react-admin should ship with everything working unrelated to my package.json.

Basically im blocked now until I have migrated my whole project to react-router v6 and all I wanted is to do a simple Button with prefilled recrod to create view.

**Additional context** 
I have already posted a bug report here related to `react-hook-form` and `useController` hook which was the same issue.

When im using a `Link` component in my `Datagrid` I get the following error:
```
Invariant failed: You should not use <Link> outside a <Router>
```

**Comment by fzaninotto** [comment]
We won't reexport the hooks from our dependencies. If react-admin uses a version of these libraries that is not compatible with the ones you use, then you can't use react-admin in your application, sorry. 

If you want a version of react-admin compatible with react-router v5, try react-admin v3.
